keps/sig-auth/20190116-service-account-external-signing.md

```diff
@@ -127,9 +159,43 @@ New token generation and validation could suffer a performance difference when r
 
 Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket.
 
-## Graduation Criteria
+## Design Details
+
+### Test Plan
+
+Unit tests covering:
+* ProjectedToken signing works with external signer client
+* OIDC keys URL shows public keys from external signer
+
+Integration tests covering:
+* ProjectedToken signing works with simple signer server
+* OIDC keys URL shows public keys from external signer
+* Validate apiserver flag logic (require issuer to be set)
+
+E2E test covering:
+* ProjectedToken signing works with simple signer server
+
+### Graduation Criteria
+
+* Alpha v1.18
+    * Start API at v1alpha1
+    * Unit test coverage of new work
+* Beta TBD
+    * Migrate API to v1beta1
+    * Integration tests added
+    * Feature flag flips to true
+    * Documentation added to website
+* Stable/GA TBD
+    * Migrate API to v1
+    * E2E tests added
+
+### Upgrade / Downgrade Strategy
+
+TODO
+
+### Version Skew Strategy
```

Comment:
We probably need to add a `version` API to allow a plugin to support multiple versions.  Thus you could upgrade your plugin to support `alpha`/`beta`/`ga` APIs at the same time and have the API server negotiate what it wants to use.  This should make upgrades/downgrades and version skews much easier to deal with.

Reply:
Does gRPC not automatically perform this via the proto package version?

Reply:
I am not familiar enough with gRPC to know what the best practice is around versioning and extending APIs over time.